File: python-services/src/financex/calculators/strategic_synthesis.py
"""strategic_synthesis — cross-signal convergence / divergence mapping.

Python pulls signals from financial_analysis (red_flags + highlights),
macro transmission, technical trend, and event impact. Classifies each
signal as positive / negative / neutral, and surfaces the groups for
the LLM to turn into a Bull/Base/Bear narrative.

No opinion is written here — only the map. The narrative lives in LLM.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
from decimal import Decimal

from pydantic import Field

from financex.schemas.analysis import FinancialAnalysisOutput
from financex.schemas.base import FinancexModel, Sector
from financex.schemas.competition import SectorComparisonReport


class Signal(FinancexModel):
    source: str           # 'fundamental' / 'macro' / 'technical' / 'event' / 'peer'
    direction: str        # 'positive' / 'negative' / 'neutral'
    label: str
    evidence: str | None = None


class SignalBucket(FinancexModel):
    positive: list[Signal] = Field(default_factory=list)
    negative: list[Signal] = Field(default_factory=list)
    neutral: list[Signal] = Field(default_factory=list)


class SynthesisOutput(FinancexModel):
    ticker: str
    period_label: str
    sector: Sector

    signals: SignalBucket = Field(default_factory=SignalBucket)
    convergence_score: Decimal = Decimal("0")      # −1..+1
    confidence: str = "low"                        # low / medium / high
    divergences: list[str] = Field(default_factory=list)

    computed_at: datetime
    schema_version: str = "1.0.0"


# ---------- Signal extraction ----------

_POSITIVE_HIGHLIGHT_CODES = {
    "GROSS_MARGIN", "EBITDA_MARGIN", "NET_MARGIN", "ROE", "ROA", "ROCE",
    "NIM", "BANK_ROE", "BANK_ROA",
    "CCC", "FCF", "ALTMAN_Z", "PIOTROSKI_F",
}
_NEGATIVE_FLAG_CODES = {
    "NET_LOSS", "LIQUIDITY_TIGHT", "OVERLEVERAGED", "INTEREST_COVERAGE_LOW",
    "ALTMAN_DISTRESS", "PIOTROSKI_WEAK",
    "BANK_NII_NEGATIVE", "BANK_COST_HIGH", "BANK_LLP_HEAVY", "BANK_ROE_WEAK",
}


def _extract_fundamental_signals(fa: FinancialAnalysisOutput) -> list[Signal]:
    out: list[Signal] = []
    # Positive highlights with healthy thresholds
    for h in fa.highlights:
        if h.code in _POSITIVE_HIGHLIGHT_CODES and h.value is not None:
            direction = "positive"
            # Direction override for ratios where lower=better
            if h.code == "CCC" and h.value > 0:
                direction = "neutral"
            if h.code == "NET_MARGIN" and h.value <= 0:
                direction = "negative"
            if h.code in {"ROE", "ROA", "BANK_ROE", "BANK_ROA"} and h.value <= 0:
                direction = "negative"
            out.append(Signal(source="fundamental", direction=direction, label=h.label, evidence=h.code))
    # Red flags map to negative
    for f in fa.red_flags:
        sev = "negative" if f.severity != "info" else "neutral"
        out.append(Signal(source="fundamental", direction=sev, label=f.code, evidence=f.message))
    return out


def _extract_peer_signals(peer_report: SectorComparisonReport | None) -> list[Signal]:
    if peer_report is None:
        return []
    out: list[Signal] = []
    for code in peer_report.strengths:
        out.append(Signal(source="peer", direction="positive", label=f"Top quartile: {code}"))
    for code in peer_report.weaknesses:
        out.append(Signal(source="peer", direction="negative", label=f"Bottom quartile: {code}"))
    return out


# ---------- Public API ----------

@dataclass(frozen=True)
class SynthesisInputs:
    financial_analysis: FinancialAnalysisOutput
    peer_report: SectorComparisonReport | None = None
    technical_trend: str | None = None    # 'bullish' / 'bearish' / 'neutral' (from ta)
    macro_tilt: str | None = None         # 'positive' / 'negative' / 'neutral'
    event_net_direction: str | None = None  # same three


def synthesize(inputs: SynthesisInputs, *, ticker: str) -> SynthesisOutput:
    fa = inputs.financial_analysis
    signals: list[Signal] = []

    signals.extend(_extract_fundamental_signals(fa))
    signals.extend(_extract_peer_signals(inputs.peer_report))

    if inputs.technical_trend:
        direction = (
            "positive" if inputs.technical_trend == "bullish"
            else "negative" if inputs.technical_trend == "bearish"
            else "neutral"
        )
        signals.append(Signal(source="technical", direction=direction, label=f"Trend: {inputs.technical_trend}"))

    if inputs.macro_tilt:
        signals.append(Signal(source="macro", direction=inputs.macro_tilt, label=f"Macro tilt: {inputs.macro_tilt}"))

    if inputs.event_net_direction:
        signals.append(Signal(
            source="event",
            direction=inputs.event_net_direction,
            label=f"Event net: {inputs.event_net_direction}",
        ))

    # Bucketize
    bucket = SignalBucket()
    for s in signals:
        if s.direction == "positive":
            bucket.positive.append(s)
        elif s.direction == "negative":
            bucket.negative.append(s)
        else:
            bucket.neutral.append(s)

    # Convergence score in [-1, 1]: (pos − neg) / total
    pos = len(bucket.positive)
    neg = len(bucket.negative)
    total = pos + neg + len(bucket.neutral)
    score = Decimal("0")
    if total > 0:
        score = ((Decimal(pos) - Decimal(neg)) / Decimal(total)).quantize(Decimal("0.01"))

    # Confidence
    if total >= 8 and abs(score) >= Decimal("0.4"):
        confidence = "high"
    elif total >= 4:
        confidence = "medium"
    else:
        confidence = "low"

    # Divergences — sources disagreeing noticeably
    divergences: list[str] = []
    sources_by_direction: dict[str, set[str]] = {"positive": set(), "negative": set()}
    for s in signals:
        if s.direction in sources_by_direction:
            sources_by_direction[s.direction].add(s.source)
    if sources_by_direction["positive"] and sources_by_direction["negative"]:
        shared = sources_by_direction["positive"] & sources_by_direction["negative"]
        for src in shared:
            divergences.append(f"{src} has both positive and negative signals — inspect closer.")

    return SynthesisOutput(
        ticker=ticker.upper(),
        period_label=fa.period_label,
        sector=fa.sector,
        signals=bucket,
        convergence_score=score,
        confidence=confidence,
        divergences=divergences,
        computed_at=datetime.now(UTC),
    )

File: backend/src/python/adapters/strategic_synthesis.ts
/**
 * strategic_synthesis adapter — cross-signal convergence map.
 *
 * Ported from python-services/src/financex/calculators/strategic_synthesis.py.
 * Extracts signed signals from the 4 deterministic upstream layers
 * (fundamental, peer, technical, macro, event), buckets them by
 * direction, and computes a convergence score in [-1, 1] plus a
 * confidence tag. The narrative (Bull/Base/Bear) stays with the LLM.
 */

type Direction = 'positive' | 'negative' | 'neutral';
type Confidence = 'low' | 'medium' | 'high';

export interface UpstreamHighlight {
  code: string;
  label?: string;
  value?: string | number | null;
  narrative_hint?: string | null;
}

export interface UpstreamRedFlag {
  code: string;
  severity: string;
  message?: string;
}

export interface UpstreamFinancialAnalysis {
  ticker?: string;
  period_label?: string;
  sector?: string;
  highlights?: UpstreamHighlight[];
  metrics?: UpstreamHighlight[];
  red_flags?: UpstreamRedFlag[];
}

export interface UpstreamSectorCompetition {
  strengths?: string[];
  weaknesses?: string[];
}

export interface UpstreamTechnical {
  trend?: string | null;
  overall_trend?: string | null;
}

export interface UpstreamMacro {
  tilt?: string | null;
  macro_tilt?: string | null;
}

export interface UpstreamEventImpact {
  portfolio_impact_summary?: string | null;
  confidence_overall?: string | null;
  event_impacts?: Array<{ impact_direction?: string | null }>;
}


export interface Signal {
  source: 'fundamental' | 'peer' | 'technical' | 'macro' | 'event';
  direction: Direction;
  label: string;
  evidence: string | null;
}


export interface LegacyStrategicSynthesisOutput {
  agent_id: string;
  output_id: string;
  ticker: string;
  period_label: string;
  sector: string;
  signals: {
    positive: Signal[];
    negative: Signal[];
    neutral: Signal[];
  };
  convergence_score: number; // −1..+1
  confidence: Confidence;
  divergences: string[];
  warnings: string[];
  review_status: string;
  source: 'python';
}


const POSITIVE_HIGHLIGHT_CODES = new Set([
  'GROSS_MARGIN', 'EBITDA_MARGIN', 'NET_MARGIN', 'ROE', 'ROA', 'ROCE',
  'NIM', 'BANK_ROE', 'BANK_ROA',
  'CCC', 'FCF', 'ALTMAN_Z', 'PIOTROSKI_F',
]);


function unwrap<T>(raw: unknown): T | null {
  if (raw == null) return null;
  if (typeof raw === 'string') {
    try { return JSON.parse(raw) as T; } catch { return null; }
  }
  return typeof raw === 'object' ? (raw as T) : null;
}


export function extractFinancialAnalysis(raw: unknown): UpstreamFinancialAnalysis | null {
  return unwrap<UpstreamFinancialAnalysis>(raw);
}

export function extractSectorCompetition(raw: unknown): UpstreamSectorCompetition | null {
  return unwrap<UpstreamSectorCompetition>(raw);
}

export function extractTechnical(raw: unknown): UpstreamTechnical | null {
  return unwrap<UpstreamTechnical>(raw);
}

export function extractMacro(raw: unknown): UpstreamMacro | null {
  return unwrap<UpstreamMacro>(raw);
}

export function extractEventImpact(raw: unknown): UpstreamEventImpact | null {
  return unwrap<UpstreamEventImpact>(raw);
}


function parseNumber(v: unknown): number | null {
  if (v == null || v === '') return null;
  const n = typeof v === 'number' ? v : Number(String(v).replace(/[, ]/g, ''));
  return Number.isFinite(n) ? n : null;
}


function fundamentalSignals(fa: UpstreamFinancialAnalysis): Signal[] {
  const out: Signal[] = [];
  const highlights = fa.highlights ?? fa.metrics ?? [];
  for (const h of highlights) {
    if (!POSITIVE_HIGHLIGHT_CODES.has(h.code)) continue;
    const val = parseNumber(h.value);
    if (val == null) continue;
    let direction: Direction = 'positive';
    if (h.code === 'CCC' && val > 0) direction = 'neutral';
    if (h.code === 'NET_MARGIN' && val <= 0) direction = 'negative';
    if (['ROE', 'ROA', 'BANK_ROE', 'BANK_ROA'].includes(h.code) && val <= 0) direction = 'negative';
    out.push({
      source: 'fundamental',
      direction,
      label: h.label ?? h.code,
      evidence: h.code,
    });
  }
  for (const f of fa.red_flags ?? []) {
    const sev = (f.severity ?? '').toLowerCase();
    const direction: Direction = sev === 'info' ? 'neutral' : 'negative';
    out.push({
      source: 'fundamental',
      direction,
      label: f.code,
      evidence: f.message ?? null,
    });
  }
  return out;
}


function peerSignals(sc: UpstreamSectorCompetition | null): Signal[] {
  if (!sc) return [];
  const out: Signal[] = [];
  for (const code of sc.strengths ?? []) {
    out.push({ source: 'peer', direction: 'positive', label: `Top quartile: ${code}`, evidence: code });
  }
  for (const code of sc.weaknesses ?? []) {
    out.push({ source: 'peer', direction: 'negative', label: `Bottom quartile: ${code}`, evidence: code });
  }
  return out;
}


function normaliseTrend(raw: string | null | undefined): Direction | null {
  if (!raw) return null;
  const v = raw.toLowerCase();
  if (v.includes('bull') || v === 'positive' || v === 'up') return 'positive';
  if (v.includes('bear') || v === 'negative' || v === 'down') return 'negative';
  if (v === 'neutral' || v.includes('side')) return 'neutral';
  return null;
}


function technicalSignal(ta: UpstreamTechnical | null): Signal | null {
  if (!ta) return null;
  const raw = ta.trend ?? ta.overall_trend;
  const dir = normaliseTrend(raw ?? null);
  if (!dir) return null;
  return { source: 'technical', direction: dir, label: `Trend: ${raw}`, evidence: null };
}


function macroSignal(ma: UpstreamMacro | null): Signal | null {
  if (!ma) return null;
  const raw = ma.tilt ?? ma.macro_tilt;
  const dir = normaliseTrend(raw ?? null);
  if (!dir) return null;
  return { source: 'macro', direction: dir, label: `Macro tilt: ${raw}`, evidence: null };
}


function eventNetSignal(ev: UpstreamEventImpact | null): Signal | null {
  if (!ev?.event_impacts?.length) return null;
  let pos = 0, neg = 0;
  for (const e of ev.event_impacts) {
    const d = (e.impact_direction ?? '').toLowerCase();
    if (d === 'positive') pos += 1;
    else if (d === 'negative') neg += 1;
  }
  let direction: Direction;
  if (pos === 0 && neg === 0) return null;
  if (pos > neg * 1.5) direction = 'positive';
  else if (neg > pos * 1.5) direction = 'negative';
  else direction = 'neutral';
  return {
    source: 'event',
    direction,
    label: `Event net: ${direction} (+${pos} / −${neg})`,
    evidence: null,
  };
}


function confidenceFrom(total: number, absScore: number): Confidence {
  if (total >= 8 && absScore >= 0.4) return 'high';
  if (total >= 4) return 'medium';
  return 'low';
}


function detectDivergences(signals: Signal[]): string[] {
  const sourcesByDir: Record<Direction, Set<string>> = {
    positive: new Set(),
    negative: new Set(),
    neutral: new Set(),
  };
  for (const s of signals) sourcesByDir[s.direction].add(s.source);
  const out: string[] = [];
  for (const src of sourcesByDir.positive) {
    if (sourcesByDir.negative.has(src)) {
      out.push(`${src} has both positive and negative signals — inspect closer.`);
    }
  }
  return out;
}


export interface SynthesisInputs {
  financialAnalysis: UpstreamFinancialAnalysis | null;
  sectorCompetition: UpstreamSectorCompetition | null;
  technical: UpstreamTechnical | null;
  macro: UpstreamMacro | null;
  eventImpact: UpstreamEventImpact | null;
}


export function adaptStrategicSynthesisForLegacy(
  inputs: SynthesisInputs,
  ticker: string,
  outputId: string,
): LegacyStrategicSynthesisOutput {
  const warnings: string[] = [];
  const fa = inputs.financialAnalysis;
  if (!fa) {
    warnings.push('No financial_analysis output — strategic_synthesis cannot extract fundamental signals');
  }

  const signals: Signal[] = [];
  if (fa) signals.push(...fundamentalSignals(fa));
  signals.push(...peerSignals(inputs.sectorCompetition));
  const ta = technicalSignal(inputs.technical); if (ta) signals.push(ta);
  const ma = macroSignal(inputs.macro);         if (ma) signals.push(ma);
  const ev = eventNetSignal(inputs.eventImpact); if (ev) signals.push(ev);

  const bucket = {
    positive: signals.filter(s => s.direction === 'positive'),
    negative: signals.filter(s => s.direction === 'negative'),
    neutral:  signals.filter(s => s.direction === 'neutral'),
  };

  const total = signals.length;
  const score = total === 0
    ? 0
    : Math.round(((bucket.positive.length - bucket.negative.length) / total) * 100) / 100;

  const confidence = confidenceFrom(total, Math.abs(score));
  const divergences = detectDivergences(signals);

  return {
    agent_id: 'strategic_synthesis',
    output_id: outputId,
    ticker: (fa?.ticker ?? ticker).toUpperCase(),
    period_label: fa?.period_label ?? 'unknown',
    sector: fa?.sector ?? 'unknown',
    signals: bucket,
    convergence_score: score,
    confidence,
    divergences,
    warnings,
    review_status: 'pending_ceo_review',
    source: 'python',
  };
}
